### ISSUE TITLE
wdis produces invalid output in favor of x87 emulation replacements

### ISSUE THREAD

**Issue by klei1984** [open-watcom/open-watcom-v2#552]
wdis version used: Open Watcom Multi-processor Disassembler Version 2.0 beta Aug 23 2019 01:43:28 (64-bit)

command line: wdis intxa386.o -l
output (intxa386.lst):

```
Segment: _TEXT WORD USE32 000003DE bytes
...
0174  CD 32				int		0x32
0176  C3				ret
0177  CD 33				int		0x33
0179  C3				ret
017A  CD 34 C3				fadd		st,st(3) ; int 0x34
017D  CD 35 C3				fld		st(3) ; int 0x35
0180  CD 36 C3				fcmovb		st,st(3) ; int 0x36
0183  CD 37 C3				fcmovnb		st,st(3) ; int 0x37
0186  CD 38 C3				fadd		st(3),st ; int 0x38
0189  CD 39 C3				ffree		st(3) ; int 0x39
018C  CD 3A C3				faddp		st(3),st ; int 0x3a
018F  CD				????
0190  3B C3				cmp		eax,ebx
0192  CD 3C C3				ret		; int 0x3c
0195  CD 3D				nop		; int 0x3d
0197  C3				ret
0198  CD 3E				int		0x3e
019A  C3				ret
...
No disassembly errors
...
```
Source file: open-watcom-v2-Current-build/bld/wdisasm/c/docode.c 
Function: LookFor87Emulation()

Problems description:
1. Int 34 to Int 3D are replaced by wdis with FPU instructions. In the above case incorrectly.
2. As LookFor87Emulation() changes the opcode size, follow up data is also incorrectly decoded. E.g.:
```
018F  CD				????
0190  3B C3				cmp		eax,ebx
```
3. Even though the actually emitted opcode is ???? the tool considers the output to be correct in the summary section: "**No disassembly errors**"

I found no option to disable LookFor87Emulation() via command line arguments.

```
#if !defined( O2A )
static int  LookFor87Emulation( void )
/************************************/

{
    char        second_byte;
    int         inssize;
    int         seg_reg;

    inssize = 0;
    if( Opcode == 0xCD ) {
        second_byte = GetNextByte();
        if( second_byte >= 0x34 && second_byte <= 0x3d ) {
            CurrIns.pref |= EMU_INTERRUPT;
            CurrIns.op[ OP_3 ].disp = second_byte;
            if( second_byte <= 0x3b ) {
                Opcode = GetDataByte() + 0xA4;
                ++inssize;
            } else if ( second_byte == 0x3D ) {
                GetDataByte();
                Opcode = 0x90;
                inssize++;
            } else {                            /* 0x3C */
                GetDataByte();
                Opcode = GetDataByte();
                seg_reg = ( ~Opcode >> 6 ) & 3;
                CurrIns.pref |= SegPrefixTab[ seg_reg ];
                CurrIns.seg_used = FIRST_SEG_REG + seg_reg;
                inssize += 2;
                Opcode |= 0xC0;
            }
        }
    }
    return( inssize );
}
#endif
```

**Comment by jmalak** [comment]
First LookFor87Emulation code doesn't exist in Open Watcom V2 disassembler.
It is code used by old disassembler (now used only by ).
Current OW disassembler is located in bld/ndisasm. 
Decoding is more complex and is spread over multiple phases (pre and post instruction processing).
In your case dissassembler is not able decode CD 3B C3 bytes sequence as valid emulated FPU instruction.
You are right that this feature can not be disabled that we should add some command line option which disable it. May be that behaviour of this feature should be changed to return to original code and decode it as INT instruction.

**Comment by jmalak** [comment]
Take into account the main purpose of disassembler core is support OW debugger for decode FPU or FPU emulator instruction.

**Comment by klei1984** [comment]
I installed the latest open-watcom-2_0-c-linux-x86 binary build on a 64 bit ubuntu linux and I only found wdis as an executable tool which is probably the old disassembler then. In the source code I found ndisasm but I doubt I have the build environment for OW. Is ndisasm supposed to be available in the binary builds by default or is it called something else and not ndisasm?

**Comment by jmalak** [comment]
The executable name is wdis for ndisasm project and it is distributed as wdis/wdis.exe.
I will fix this problem, but there are other logical problems.
- for 32-bit code interrupt 0x34-0x3D are not used for FPU emulation that it is always interrupt
- for 16-bit code disassembly it should be handled by fixupp, if no fixupp then it is interrupt
- for debugger (MAD) it should be fixed for 32-bit code (no emulation, always interrupts)

**Comment by jmalak** [comment]
it is fixed now by 60e35020738b2132b1a9b9728904a647c5691582 commit

**Comment by klei1984** [comment]
I can confirm that the disassembly issues I reported are now gone.
